### PR TITLE
Respect DOTNET_SHUTDOWNTIMEOUTSECONDS in HostOptions

### LINKED ISSUE #36059: DOTNET_SHUTDOWNTIMEOUTSECONDS not respected
### Setting ShutdownTimeout via environment variable does not affect host shutdown timeout
According to the [docs](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/host/generic-host?view=aspnetcore-3.1#shutdowntimeout), setting the DOTNET_SHUTDOWNTIMEOUTSECONDS as an environment variable should affect graceful shutdown time. Configuring it through code work fine.

### To Reproduce
1. Create new api project from VS template
2. Set the environment variable in debug settings (or otherwise), to a value other than the 5 seconds default.
3. Add an IOptions< HostOptions > dependency to the default created controller and return the ShutdownTimeout.
4. Start in debug.
5. In the browser, see that the value is still 5 seconds.

### Expected behavior
The HostOptions.ShutdownTimeout value should be set to whatever value passed to it by DOTNET_SHUTDOWNTIMEOUTSECONDS.

Tested both with 3.1 and 3.0.1.

Please see attached repro.
[ReproduceShutdownTimeout.zip](https://github.com/dotnet/extensions/files/4164843/ReproduceShutdownTimeout.zip)

### PR DESCRIPTION
**Pull request by galakt**
Fixes #36059

### CODE REVIEW

File: src/libraries/Microsoft.Extensions.Hosting/src/HostOptions.cs [hunk 1]
```diff
@@ -10,6 +11,16 @@ namespace Microsoft.Extensions.Hosting
     /// </summary>
     public class HostOptions
     {
+        public HostOptions()
+        {
+            var timeoutSeconds = Environment.GetEnvironmentVariable("DOTNET_SHUTDOWNTIMEOUTSECONDS");
+            if (!string.IsNullOrWhiteSpace(timeoutSeconds)
+                && int.TryParse(timeoutSeconds, NumberStyles.None, CultureInfo.InvariantCulture, out var seconds))
+            {
+                ShutdownTimeout = TimeSpan.FromSeconds(seconds);
```
Comment (maryamariyan) [inline src/libraries/Microsoft.Extensions.Hosting/src/HostOptions.cs:20]: Test below was manually setting `ShutdownTimeout`

https://github.com/dotnet/runtime/blob/ae908fec2549ae2e901124794faf2f2233461ec8/src/libraries/Microsoft.Extensions.Hosting/tests/UnitTests/Internal/HostTests.cs#L599-L629

Let's get a test covering the case you are adding here.

Comment (galakt) [reply src/libraries/Microsoft.Extensions.Hosting/src/HostOptions.cs:20]: But how I supposed to test my case if I get value directly from Env, not from `IConfiguration`? I mean I see no way to mock Environment

Comment (maryamariyan) [reply src/libraries/Microsoft.Extensions.Hosting/src/HostOptions.cs:20]: you're right. let me check.

@HaoK do you have a recommendation for this?

Comment (maryamariyan) [reply src/libraries/Microsoft.Extensions.Hosting/src/HostOptions.cs:20]: I have to first investigate how this is handled this for "ASPNET_SHUTDOWNTIMEOUTSECONDS".

I think the approach needs to change a bit not just the test.

Comment (galakt) [reply src/libraries/Microsoft.Extensions.Hosting/src/HostOptions.cs:20]: "ASPNET_SHUTDOWNTIMEOUTSECONDS" does not work for Generic Host, created separated issue #44101

File: src/libraries/Microsoft.Extensions.Hosting/src/HostOptions.cs [hunk 2]
```diff
@@ -10,6 +11,16 @@ namespace Microsoft.Extensions.Hosting
     /// </summary>
     public class HostOptions
     {
+        public HostOptions()
+        {
+            var timeoutSeconds = Environment.GetEnvironmentVariable("DOTNET_SHUTDOWNTIMEOUTSECONDS");
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Hosting/src/HostOptions.cs:16]: I guess I expected the fix here was to inject an `IConfiguration` into the HostOptions class - just like the WebHost does. Is there a reason we aren't taking that approach?

https://github.com/dotnet/aspnetcore/blob/fd1f1c355984f76070d8e32c891e6bac0bd96fdf/src/Hosting/Hosting/src/Internal/WebHostOptions.cs#L19-L40

Comment (maryamariyan) [reply src/libraries/Microsoft.Extensions.Hosting/src/HostOptions.cs:16]: Yes, I also think we shouldn't need to use GetEnvironmentVariable here. then the testing can be done using IConfiguration as well

Comment (galakt) [reply src/libraries/Microsoft.Extensions.Hosting/src/HostOptions.cs:16]: @eerhardt Options must have parameterless constructor, so `HostOptions` will have two constructors (Changes to this file must follow the https://aka.ms/api-review process). But we can not just get `HostOptions` with `IConfiguration` from DI when we have parameterless constructor, can we?

Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.Hosting/src/HostOptions.cs:16]: DI will pick the constructor with the longest signature. So while we will keep the parameter less ctor for anyone creating this class manually, DI will always create it using the IConfiguration.

Comment (galakt) [reply src/libraries/Microsoft.Extensions.Hosting/src/HostOptions.cs:16]: @eerhardt afaik this rule does not work for IOptions injection https://github.com/dotnet/runtime/blob/0eb6c4eba8b391db117a22bf60d9eaec9695fa97/src/libraries/Microsoft.Extensions.Options/src/IOptions.cs#L12-L19
https://github.com/dotnet/runtime/blob/0eb6c4eba8b391db117a22bf60d9eaec9695fa97/src/libraries/Microsoft.Extensions.Options/src/Options.cs#L11-L14

Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.Hosting/src/HostOptions.cs:16]: Ah, you are right - I was missing that it is the `OptionsMonitor` that gets injected in DI, and not the Options object itself. Looking at ASP.NET, they don't use DI at all to create the WebHostOptions object - they do it themselves.

One way to solve this issue is to use a `.Configure<HostOptions>` call to hook the configuration up to the options. Here's an example:

https://github.com/eerhardt/runtime/commit/01ce98195405b51080e5a80e53902abe1f3fbe0b

Comment (galakt) [reply src/libraries/Microsoft.Extensions.Hosting/src/HostOptions.cs]: Updated

**Review by eerhardt — APPROVED** [review]
LGTM. Thanks for the contribution, @galakt!

**Review by maryamariyan — APPROVED** [review]

**Pull request merged**